File: 07-trees/00098-Validate_Binary_Search_Tree/00098-validate_binary_search_tree.go
// 98: Validate Binary Search Tree
// https://leetcode.com/problems/validate-binary-search-tree/

package main

import (
	"fmt"
)

type TreeNode struct {
    val any
    left *TreeNode
    right *TreeNode
}

type Tree struct {
    root *TreeNode
}

func creator(values []any, root **TreeNode, i, n int) *TreeNode {
    if n==0 {return nil}
    if i<n {
        temp := &TreeNode{values[i], nil, nil}
        *root = temp
        (*root).left = creator(values, &((*root).left), 2*i+1, n);
        (*root).right = creator(values, &((*root).right), 2*i+2, n);
    }
    return *root;
}

func createTree(root **TreeNode, inputs []any) {
    creator(inputs, root, 0, len(inputs))
}

func showTree(root *TreeNode) {
    var q []*TreeNode;
    var result [][]interface{}
    var c []interface{}
    if root==nil { fmt.Println("Empty !"); return; }
    q = append(q, root)
    q = append(q, nil)
    for len(q)!=0 {
        t := q[0]
        q = q[1:]
        if t==nil {
            result = append(result, c)
            c = make([]interface{}, 0)
            if len(q) > 0 {q = append(q, nil)}
        } else {
            c = append(c, t.val)
            if t.left!=nil {q = append(q, t.left)}
            if t.right!=nil {q = append(q, t.right)}
        }
    }

    fmt.Print("["); for _, x := range result {
        fmt.Print("["); for _, y := range x {
            if y==nil { fmt.Print("NULL,"); continue; }
            fmt.Print(y,",")
        }; fmt.Print("\b],")
    }; fmt.Println("\b]")
}

// SOLUTION
func traverse(root, min, max *TreeNode) bool {
	if root==nil {return true}
	if min!=nil && root.val.(int)<=min.val.(int) {return false}
	if max!=nil && root.val.(int)>=max.val.(int) {return false}

	l := traverse(root.left, min, root)
	r := traverse(root.right, root, max)

	return l && r
}

func isValidBST(root *TreeNode) bool {
    return traverse(root, nil, nil)
}


func main() {
    tree := Tree{}

    // INPUT
    tn := []any{2,1,3}
    createTree(&tree.root, tn)

    // OUTPUT
    result := isValidBST(tree.root)
	fmt.Println(result)
}

File: 07-trees/00098-Validate_Binary_Search_Tree/00098-validate_binary_search_tree.java
// 98: Validate Binary Search Tree
// https://leetcode.com/problems/validate-binary-search-tree/



import java.util.Queue;
import java.util.ArrayList;
import java.util.LinkedList;


class Solution {
    public static class TreeNode {
        Integer val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(Integer val) { this.val = val; }
        TreeNode(Integer val, TreeNode left, TreeNode right) { this.val = val; this.left = left; this.right = right; }
    }

    public static class Tree {
        TreeNode root;
        Tree() { this.root = null; }
    }

    private static TreeNode creator(Integer[] values, TreeNode root, int i, int n) {
        if (n==0) return null;
        if (i<n) {
            TreeNode temp = new TreeNode(values[i]);
            root = temp;
            root.left = creator(values, root.left, 2*i+1, n);
            root.right = creator(values, root.right, 2*i+2, n);
        }
        return root;
    }

    public static void createTree(Tree tree, Integer[] inputs) {
        tree.root = creator(inputs, tree.root, 0, inputs.length);
    }

    public static void showTree(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        ArrayList<ArrayList<Integer>> result = new ArrayList<>();
        ArrayList<Integer> c = new ArrayList<>();
        if (root == null) { System.out.println("Empty !"); return; }
        q.add(root);
        q.add(null);
        while (!q.isEmpty()) {
            TreeNode t = q.poll();
            if (t==null) {
                result.add(c);
                c = new ArrayList<>();
                if (q.size() > 0) q.add(null);
            } else {
                c.add(t.val);
                if (t.left!=null) q.add(t.left);
                if (t.right!=null) q.add(t.right);
            }
        }

        System.out.print("["); for (var x : result) {
            System.out.print("["); for (var y : x) {
                if (y==null) { System.out.print("NULL,"); continue; }
                System.out.print(y+",");
            } System.out.print("\b],");
        } System.out.println("\b]");
    }


    // SOLUTION    
    private boolean traverse(TreeNode root, TreeNode min, TreeNode max) {
        if (root==null) return true;
        if (min!=null && root.val<=min.val) return false;
        if (max!=null && root.val>=max.val) return false;

        boolean l = traverse(root.left, min, root);
        boolean r = traverse(root.right, root, max);

        return l && r;
    }

    public boolean isValidBST(TreeNode root) {
        return traverse(root, null, null);
    }


    public static void main(String[] args) {
        Solution o = new Solution();
        Tree tree = new Tree();

        // INPUT
        Integer[] tn = {2,1,3};
        createTree(tree, tn);

        // OUTPUT
        var result = o.isValidBST(tree.root);
        System.out.println(result);
    }
}
